File: tests/oracle/cspec_global_db_1204.cc
/*
 * CSPEC-GLOBAL-DB-0001: locked Ghidra 12.0.4 global-scope Database
 * write-path oracle (CSPEC-GLOBAL-APPLY-0001 B2 fixture).  Loads the real
 * production x86-64-gcc.cspec through a full BfdArchitecture::init (which
 * drives Architecture::parseCompilerConfig -> decodeGlobal ->
 * addToGlobalScope -> Database::addRange), then prints the Database-side
 * observations the Rust fixture must reproduce byte for byte from the
 * same cspec bytes:
 *   - DBTREE: the global scope's space-keyed ownership tree through the
 *     public Scope::printBounds (set<Range> sorted by (space index,
 *     first), one "<space>: <first hex>-<last hex>" line per range,
 *     address.cc:283/588),
 *   - QPROP: Scope::queryProperties flag folds (database.cc:1263-1281)
 *     at representative (space, offset, size) probes —
 *     ram inside the <global> range (mapped|addrtied|persist),
 *     register inside the cspec register window (same fold),
 *     register/unique outside any owned range (property-only), OTHER
 *     inside its whole-space range (fold), const (stackContainer's
 *     cc:950 isConstant bail -> property-only).
 *
 * Probe offsets in ram are confined to the pinned curl binary's sole
 * PF_W segment (vaddr 0x16c48..0x18680): the loader readonly property
 * ranges (fillinReadOnlyFromLoader) live in the R/RX segments only, so
 * both sides fold getProperty()==0 there and the observation stays on
 * the global-range fold this fixture owns.
 */
#include "bfd_arch.hh"
#include "fspec.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "xml.hh"

#include <fstream>
#include <iostream>
#include <list>
#include <sstream>
#include <stdexcept>
#include <string>
#include <cstdlib>
#include <cstring>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <vector>

namespace {

using namespace ghidra;
using std::cerr;
using std::cout;
using std::exception;
using std::getline;
using std::istringstream;
using std::ostringstream;
using std::runtime_error;
using std::string;
using std::vector;

void copyFile(const string &from, const string &to)
{
  ifstream in(from.c_str(), std::ios::binary);
  if (!in)
    throw runtime_error("cannot open " + from);
  ofstream out(to.c_str(), std::ios::binary);
  out << in.rdbuf();
  if (!out)
    throw runtime_error("cannot write " + to);
}

// Private working copy of the spec set (never mutates the input
// directory): specpaths.findFile resolves by first-registered directory.
string makeWorkSpecDir(const string &specDirectory)
{
  char templatePath[] = "/tmp/cspec-global-db-work.XXXXXX";
  const char *created = ::mkdtemp(templatePath);
  if (created == (const char *)0)
    throw runtime_error("mkdtemp failed");
  const string dir(created);
  const char *names[] = {"x86.ldefs", "x86-64.pspec", "x86-64.sla", "x86-64-gcc.cspec"};
  for (int i = 0; i < 4; ++i)
    copyFile(specDirectory + "/" + names[i], dir + "/" + names[i]);
  return dir;
}

struct Probe
{
  const char *spaceName;
  uintb offset;
  int4 size;
};

void runFixture(const string &specDirectory, const string &binary)
{
  const string workDir = makeWorkSpecDir(specDirectory);
  vector<string> specPaths;
  specPaths.push_back(workDir);
  startDecompilerLibrary(specPaths);
  BfdArchitecture arch(binary, "default", &std::cerr);
  DocumentStorage documents;
  arch.init(documents);

  cout << "SCHEMA|1\n";

  // The global scope's ownership tree: Scope::printBounds (database.hh:789
  // -> RangeList::printBounds address.cc:588, per-Range address.cc:283).
  {
    const Scope *globalScope = arch.symboltab->getGlobalScope();
    ostringstream buffer;
    globalScope->printBounds(buffer);
    istringstream lines(buffer.str());
    string line;
    while (getline(lines, line)) {
      if (line == "all")
        continue;
      cout << "DBTREE|" << line << "\n";
    }
  }

  // queryProperties folds at the probe grid. flags print as 0x%08x so the
  // Varnode bit positions (addrtied=0x8000, persist=0x4000,
  // mapped=0x200000) are directly visible; the usepoint is the invalid
  // Address() (the newVarnode call-site form, funcdata_varnode.cc:148-166).
  static const Probe probes[] = {
    {"ram", 0x17000, 8},
    {"ram", 0x18000, 1},
    {"register", 0x1094, 4},
    {"register", 0x1080, 2},
    {"register", 0x1096, 2},
    {"unique", 0x100, 4},
    {"OTHER", 0x10, 1},
    {"const", 0x5, 1},
  };
  const Scope *globalScope = arch.symboltab->getGlobalScope();
  for (int4 i = 0; i < 8; ++i) {
    const Probe &probe = probes[i];
    AddrSpace *spc = arch.getSpaceByName(probe.spaceName);
    if (spc == (AddrSpace *)0)
      throw runtime_error(string("missing space: ") + probe.spaceName);
    Address addr(spc, probe.offset);
    Address usepoint;
    uint4 flags = 0;
    SymbolEntry *entry = globalScope->queryProperties(addr, probe.size, usepoint, flags);
    cout << "QPROP|" << probe.spaceName << "|0x" << std::hex << probe.offset
         << "|" << std::dec << probe.size << "|0x" << std::hex << flags
         << "|" << std::dec << (entry == (SymbolEntry *)0 ? 0 : 1) << "\n";
  }
}

} // namespace

int main(int argc, char **argv)
{
  if (argc != 3) {
    cerr << "usage: cspec_global_db_1204 <spec-directory> <binary>\n";
    return 2;
  }
  try {
    runFixture(argv[1], argv[2]);
    return 0;
  }
  catch (LowlevelError &error) {
    cerr << error.explain << '\n';
  }
  catch (DecoderError &error) {
    cerr << error.explain << '\n';
  }
  catch (exception &error) {
    cerr << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/cspec_global_db_1204.rs
//! CSPEC-GLOBAL-DB-0001: Rust side of the locked Ghidra 12.0.4
//! global-scope Database write-path oracle (CSPEC-GLOBAL-APPLY-0001 B2
//! fixture).  Ingests the same production x86-64-gcc.cspec bytes through
//! the real marshal `DocumentStorage` text parser, drives
//! `Architecture::parse_compiler_config` (whose `add_to_global_scope` /
//! `add_other_space` write the cspec `<global>` + OTHER triples into the
//! constructor symbol table's global scope), and prints the same
//! Database-side observations the C++ fixture prints
//! (tests/oracle/cspec_global_db_1204.cc) so the runner can diff the two
//! byte for byte:
//!   - DBTREE: the global scope's space-keyed ownership tree
//!     (`Scope::printBounds`, address.cc:283/588 form),
//!   - QPROP: `Database::query_properties_spaced` folds at the same
//!     (space, offset, size) probe grid (database.cc:1263-1281).

use rugra::arch::{Architecture, SpecQuery};
use rugra::fspec::VarnodeData;
use rugra::marshal::DocumentStorage;
use rugra::sleigh_ffi::{set_sla_path, SleighCtx};
use rugra::space::AddressSpace;

use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::io::Write as _;
use std::process::ExitCode;
use std::sync::{Arc, RwLock};

/// Locked x86-64 address-space facts (the same table the text-ingest
/// fixture mirrors): name → highest offset.
const SPACE_HIGHEST: [(&str, u64); 9] = [
    ("const", u64::MAX),
    ("OTHER", u64::MAX),
    ("unique", 0xffff_ffff),
    ("ram", u64::MAX),
    ("register", 0xffff_ffff),
    ("fspec", u64::MAX),
    ("iop", u64::MAX),
    ("join", 0xffff_ffff),
    ("stack", u64::MAX),
];

/// Unique-space inject base (locked x86-64 .sla fact).
const UNIQUE_INJECT_BASE: u64 = 0x364_400;

fn space_of(name: &str) -> Option<AddressSpace> {
    match name {
        "ram" => Some(AddressSpace::Ram),
        "stack" => Some(AddressSpace::Stack),
        "register" => Some(AddressSpace::Register),
        "OTHER" | "other" => Some(AddressSpace::Other(1)),
        "unique" => Some(AddressSpace::Unique),
        "const" => Some(AddressSpace::Const),
        _ => None,
    }
}

struct Host {
    registers: BTreeMap<String, VarnodeData>,
}

impl rugra::pcodeparse::SleighSymbolLookup for Host {
    fn find_symbol(&self, name: &str) -> Option<rugra::pcodeparse::SleighSymbol> {
        self.registers.get(name).map(|vd| rugra::pcodeparse::SleighSymbol {
            name: name.to_string(),
            kind: rugra::pcodeparse::SleightSymbolKind::Varnode(rugra::varnode::VarnodeData {
                space: vd.space,
                offset: vd.offset,
                size: vd.size.max(0) as usize,
            }),
        })
    }
}

impl SpecQuery for Host {
    fn get_register(&self, name: &str) -> Option<VarnodeData> {
        self.registers.get(name).copied()
    }
    fn space_by_name(&self, name: &str) -> Option<AddressSpace> {
        space_of(name)
    }
    fn space_highest(&self, spc: AddressSpace) -> u64 {
        let name = match spc {
            AddressSpace::Ram => "ram",
            AddressSpace::Stack => "stack",
            AddressSpace::Register => "register",
            AddressSpace::Other(_) => "OTHER",
            AddressSpace::Unique => "unique",
            AddressSpace::Const => "const",
            AddressSpace::Iop => "iop",
            AddressSpace::Join => "join",
            AddressSpace::Overlay => "overlay",
        };
        SPACE_HIGHEST
            .iter()
            .find(|(n, _)| *n == name)
            .map(|(_, h)| *h)
            .unwrap_or(u64::MAX)
    }
    fn unique_inject_base(&self) -> u64 {
        UNIQUE_INJECT_BASE
    }
}

fn run() -> Result<(), String> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        return Err("usage: cspec_global_db_1204 <cspec> <sla>".to_string());
    }
    let cspec_bytes =
        fs::read(&args[1]).map_err(|e| format!("failed to read {}: {}", args[1], e))?;
    let mut store = DocumentStorage::new();
    let doc = store
        .parse_document(&cspec_bytes)
        .map_err(|e| format!("cspec parse failed: {}", e))?;
    let root = doc
        .root
        .clone()
        .ok_or_else(|| "cspec has no root element".to_string())?;
    if root.read().map_err(|_| "poisoned lock")?.name != "compiler_spec" {
        return Err("cspec root is not compiler_spec".to_string());
    }
    store.register_tag(&root);

    // Registers from the real .sla (the cspec register range resolves
    // through the register table, architecture.cc:829-831).
    set_sla_path(&args[2]);
    let sleigh = SleighCtx::new().ok_or_else(|| "SleighCtx::new failed".to_string())?;
    let mut registers = BTreeMap::new();
    for index in 0..sleigh.num_registers() {
        if let Some((name, space, offset, size)) = sleigh.register_info(index) {
            let Ok(space_id) = u8::try_from(space) else {
                continue;
            };
            registers.insert(
                name,
                VarnodeData {
                    space: AddressSpace::from_id(space_id),
                    offset,
                    size,
                },
            );
        }
    }
    let host = Arc::new(Host { registers });

    // The Architecture with the injection library (SLEIGH lookup installed
    // like PcodeInjectLibrarySleigh's slgh member) — the <callfixup>
    // snippets compile through it during parse.
    let mut arch = Architecture::new();
    arch.archid = "x86:LE:64:default".to_string();
    let mut inject_lib = rugra::pcodeinject::PcodeInjectLibrary::new(UNIQUE_INJECT_BASE);
    inject_lib.set_sleigh_lookup(host.clone());
    arch.pcodeinjectlib = Some(Arc::new(RwLock::new(inject_lib)));
    let mut userops = rugra::userop::UserOpManage::new();
    userops.register_op("segment".to_string(), rugra::userop::UserOpType::Unspecialized);
    arch.userops = Some(Arc::new(RwLock::new(userops)));
    arch.parse_compiler_config(&mut store, host.as_ref(), 8)
        .map_err(|e| format!("parse_compiler_config failed: {}", e))?;

    let mut out = String::new();
    out.push_str("SCHEMA|1\n");

    let symboltab = arch
        .symboltab
        .as_ref()
        .ok_or("constructor symboltab missing")?
        .clone();
    let db = symboltab.read().map_err(|_| "poisoned lock")?;
    let global_scope_id = db.global_scope_id;

    // DBTREE: the global scope's ownership tree through the public
    // printBounds (address.cc:588 form, `all` suppressed like the C++
    // fixture's line filter).
    {
        let global = db
            .get_global_scope()
            .ok_or("global scope missing")?;
        let bounds = global.print_bounds();
        for line in bounds.lines() {
            if line == "all" {
                continue;
            }
            out.push_str(&format!("DBTREE|{}\n", line));
        }
    }

    // QPROP: the same probe grid as the C++ fixture. Ram probes sit in
    // the pinned curl binary's sole PF_W segment so no loader-readonly
    // property participates on either side.
    let probes: [(&str, u64, i32); 8] = [
        ("ram", 0x17000, 8),
        ("ram", 0x18000, 1),
        ("register", 0x1094, 4),
        ("register", 0x1080, 2),
        ("register", 0x1096, 2),
        ("unique", 0x100, 4),
        ("OTHER", 0x10, 1),
        ("const", 0x5, 1),
    ];
    for (space_name, offset, size) in probes {
        let spc = space_of(space_name).ok_or("unknown probe space")?;
        let (hit, flags) = db.query_properties_spaced(
            global_scope_id,
            spc,
            offset,
            size,
            rugra::address::Address::new(0), // the invalid-usepoint form
        );
        out.push_str(&format!(
            "QPROP|{}|0x{:x}|{}|0x{:x}|{}\n",
            space_name,
            offset,
            size,
            flags,
            if hit.is_some() { 1 } else { 0 }
        ));
    }

    let mut stdout = std::io::stdout();
    stdout
        .write_all(out.as_bytes())
        .map_err(|e| format!("stdout write failed: {}", e))?;
    stdout.flush().map_err(|e| format!("stdout flush failed: {}", e))?;
    Ok(())
}

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("{}", error);
            ExitCode::FAILURE
        }
    }
}
